### PR TITLE
Cassandra 15901

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
- [CI j11](https://app.circleci.com/pipelines/github/bereng/cassandra/52/workflows/573ad5be-e34d-4668-a0af-2726d4b35568) Failure seems unrelated
- [CI j8](https://app.circleci.com/pipelines/github/bereng/cassandra/52/workflows/16e15155-7dce-4877-86f5-315c6a837d36) Seems to be a new flaky tests but unrelated to the PR imo
- Run on [cassandra13](https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-test/156/) with some flakyness
- Run on [another node](https://ci-cassandra.apache.org/job/Cassandra-devbranch-test/157/) is stable

File: src/java/org/apache/cassandra/utils/FBUtilities.java [hunk 1]
```diff
@@ -130,16 +130,24 @@ public static MessageDigest newMessageDigest(String algorithm)
...
+            if (DatabaseDescriptor.getListenAddress() == null)
             {
-                throw new RuntimeException(e);
+                try
+                {
+                    localInetAddress = InetAddress.getLocalHost();
+                }
+                catch(UnknownHostException e)
+                {
+                    logger.info("InetAddress.getLocalHost() could not resolve, please check your node's config. Falling back to {}",
```
Comment (snazy) [inline src/java/org/apache/cassandra/utils/FBUtilities.java]:
```suggestion
                    logger.info("InetAddress.getLocalHost() could not resolve the address for the hostname ({}), please check your node's config and set the listen_address in cassandra.yaml. Falling back to {}",
```

File: src/java/org/apache/cassandra/utils/FBUtilities.java [hunk 2]
```diff
@@ -130,16 +130,24 @@ public static MessageDigest newMessageDigest(String algorithm)
...
+                                InetAddress.getLoopbackAddress());
```
Comment (snazy) [inline src/java/org/apache/cassandra/utils/FBUtilities.java:148]:
```suggestion
                               e, InetAddress.getLoopbackAddress());
```

File: src/java/org/apache/cassandra/utils/FBUtilities.java [hunk 3]
```diff
@@ -130,16 +130,24 @@ public static MessageDigest newMessageDigest(String algorithm)
...
-                                    : DatabaseDescriptor.getListenAddress();
-            }
-            catch (UnknownHostException e)
+        {
+            if (DatabaseDescriptor.getListenAddress() == null)
             {
-                throw new RuntimeException(e);
+                try
+                {
+                    localInetAddress = InetAddress.getLocalHost();
```
Comment (snazy) [inline src/java/org/apache/cassandra/utils/FBUtilities.java:138]:
```suggestion
                    localInetAddress = InetAddress.getLocalHost();
                    logger.info("InetAddress.getLocalHost() was used to resolve listen_address to {}, this can cause issues. Please check your node's config and set the listen_address in cassandra.yaml.", localInetAddress);
```

File: conf/cassandra.yaml [hunk 4]
```diff
@@ -630,7 +630,8 @@ ssl_storage_port: 7001
 # Leaving it blank leaves it up to InetAddress.getLocalHost(). This
 # will always do the Right Thing _if_ the node is properly configured
 # (hostname, name resolution, etc), and the Right Thing is to use the
-# address associated with the hostname (it might not be).
+# address associated with the hostname (it might not be). If unresolvable
+# it will get InetAddress.getLoopbackAddress()
```
Comment (snazy) [inline conf/cassandra.yaml]:
```suggestion
# it will fall back to InetAddress.getLoopbackAddress(), which is wrong for production systems.
```

**Review by snazy — COMMENTED** [review]
LGTM but I left a few suggestions around wording and log messages. WDYT?

**Review by michaelsembwever — APPROVED** [review]

File: src/java/org/apache/cassandra/utils/FBUtilities.java (hunk 3 above)
Comment (bereng) [reply src/java/org/apache/cassandra/utils/FBUtilities.java:138]: @snazy I would reword

> InetAddress.getLocalHost() was used to resolve listen_address to {}, this can cause issues. Please check your node's config and set the listen_address in cassandra.yaml

to

> InetAddress.getLocalHost() was used to resolve listen_address to {}, double check this is correct. Please check your node's config and set the listen_address in cassandra.yaml accordingly if applicable.

Why? getLocalHost may be the right thing and `can cause issues` sounds like almost a certainty it will cause them. I'd favor a milder wording. Also listen_address you only set up unless you set a device to listen on iirc they are exclusive, so you set listen_address only if applicable. Wdyt?

Comment (snazy) [reply src/java/org/apache/cassandra/utils/FBUtilities.java:138]: Yea, not opposed to a milder wording as long as it becomes clear that people should always configure it and not rely on (dangerous) defaults.

**Comment by bereng** [comment]
CI runs:

[j8](https://app.circleci.com/pipelines/github/bereng/cassandra/59/workflows/44042930-ea9f-4b02-88d6-91b36d1d9bd6):
- j8unit ActiveRepairServiceTest: Passes locally, ConnectionTest: [Known](https://ci-cassandra.apache.org/job/Cassandra-trunk/202/testReport/org.apache.cassandra.net/ConnectionTest/testMessagePurging/) offender
- j8_dtests-no-vnodes TestTransientReplicationRing looks [flaky](https://ci-cassandra.apache.org/job/Cassandra-trunk/202/testReport/dtest-novnode.transient_replication_ring_test/TestTransientReplicationRing/test_remove/history/) and passes locally
- j8_dtests-with-vnodes TestIncRepair [known](https://issues.apache.org/jira/browse/CASSANDRA-15894) offender
- j11unit ValidationExecutorTest passes locally
- j11 dtests-with-vnodes test_manual_session_force_cancel passes locally, [test_restart_node_localhost](https://ci-cassandra.apache.org/job/Cassandra-trunk/202/testReport/dtest.pushed_notifications_test/TestPushedNotifications/test_restart_node_localhost/) known offender

[j11](https://app.circleci.com/pipelines/github/bereng/cassandra/59/workflows/f144b763-38e8-4fd5-b380-eaa6df88a59a)
- j11unit testNoTreesRetainedAfterDifference passes locally
- j11jvm_dtest Weird VM [failure](https://app.circleci.com/pipelines/github/bereng/cassandra/59/workflows/f144b763-38e8-4fd5-b380-eaa6df88a59a/jobs/544) and passes locally
- dtests-no-vnodes: test_repaired_tracking_with_partition_deletes passes locally

Ci-cass juni one [weird](https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch-test/165/) VM failure, unrelated?

[ci-cass](https://ci-cassandra.apache.org/view/patches/job/Cassandra-devbranch/196/) full run. Note: jenkins final report is broken, it collects results from other runs, check stages individually making sure commit ids match.
- [ChecksummingTransformerTest](https://ci-cassandra.apache.org/job/Cassandra-devbranch-test-cdc/126/testReport/junit/org.apache.cassandra.transport.frame.checksum/ChecksummingTransformerTest/corruptionCausesFailure_cdc/) I have seen this fail locally and looks [flaky](https://ci-cassandra.apache.org/job/Cassandra-trunk/202/testReport/org.apache.cassandra.transport.frame.checksum/ChecksummingTransformerTest/history/)
- [LongSharedExecutorPoolTest.testPromptnessOfExecution](https://ci-cassandra.apache.org/job/Cassandra-devbranch-test-burn/131/testReport/junit/org.apache.cassandra.concurrent/LongSharedExecutorPoolTest/testPromptnessOfExecution/) passes locally
- [testMessagePurging-compression](https://ci-cassandra.apache.org/job/Cassandra-devbranch-test-compression/128/testReport/junit/org.apache.cassandra.net/ConnectionTest/testMessagePurging_compression/) known [offender](https://ci-cassandra.apache.org/job/Cassandra-trunk/202/testReport/org.apache.cassandra.net/ConnectionTest/testMessagePurging/history/) also locally
- TestPushedNotifications.test_restart_node_localhost is a known failure [CASSANDRA-15912](https://issues.apache.org/jira/browse/CASSANDRA-15912)
